include/seastar/util/chunked_deque.hh

```diff
@@ -0,0 +1,733 @@
+    } catch(...) {
+        undo_room_front();
+        throw;
+    }
+    _front_chunk->begin = next;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+template <typename... Args>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::emplace_back(Args&&... args) {
+    ensure_room_back();
+    auto p = &_back_chunk->items[_back_chunk->end].data;
+    try {
+        new(p) T(std::forward<Args>(args)...);
+    } catch(...) {
+        undo_room_back();
+        throw;
+    }
+    ++_back_chunk->end;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::push_back(const T& data) {
+    ensure_room_back();
+    auto p = &_back_chunk->items[_back_chunk->end].data;
+    try {
+        new(p) T(data);
+    } catch(...) {
+        undo_room_back();
+        throw;
+    }
+    ++_back_chunk->end;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline void
+chunked_deque<T, items_per_chunk, save_free_chunks>::push_back(T&& data) {
+    ensure_room_back();
+    auto p = &_back_chunk->items[_back_chunk->end].data;
+    try {
+        new(p) T(std::move(data));
+    } catch(...) {
+        undo_room_back();
+        throw;
+    }
+    ++_back_chunk->end;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline T&
+chunked_deque<T, items_per_chunk, save_free_chunks>::front() noexcept {
+    return _front_chunk->items[_front_chunk->begin].data;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline const T&
+chunked_deque<T, items_per_chunk, save_free_chunks>::front() const noexcept {
+    return _front_chunk->items[_front_chunk->begin].data;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline
+T&
+chunked_deque<T, items_per_chunk, save_free_chunks>::back() noexcept {
+    return _back_chunk->items[_back_chunk->end - 1].data;
+}
+
+template <typename T, size_t items_per_chunk, int save_free_chunks>
+inline
+const T&
+chunked_deque<T, items_per_chunk, save_free_chunks>::back() const noexcept {
+    return _back_chunk->items[_back_chunk->end - 1].data;
+}
+
+// Certain use cases may need to repeatedly allocate and free a chunk -
+// an obvious example is an empty queue to which we push, and then pop,
+// repeatedly. Another example is pushing and popping to a non-empty deque
+// we push and pop at different chunks so we need to free and allocate a
+// chunk every items_per_chunk operations.
+// The solution is to keep a list of freed chunks instead of freeing them
+// immediately. There is a performance/memory tradeoff of how many freed
+// chunks to save: If we save them all, the queue can never shrink from
+// its maximum memory use (this is how circular_buffer behaves).
+// The ad-hoc choice made here is to limit the number of saved chunks to 1,
+// but this could easily be made a configuration option.
```

Comment:
It seems you made save_free_chunks into a configuration option, so you can remove the last two lines of the comment above.

Reply:
thanks, will do.